Fixes #12030 subscription data not getting updated after candlepin mo…

johnpmitsch opened the pull request:
…difies it

Left to do
- [x] refactor and update rabl templates to return content host subscriptions directly from cp
- [x] show guest/temporary subscriptions

This also fixes #12188 and #12029

johnpmitsch commented:
Here the quantity information for a content host or activation key subscription is recieved from candlepin and passed through the API to the UI. This is done because the subscription/pool in the AR database will not have information about the quantity attached to a particular content host or ack key.

This fixes some main issues with the recent subscription refactoring, but there is still work left to do, notably a refactoring of content hosts to show guest subscriptions.

What I am thinking will be best to solve the remaining issues and make things easier is create a new, duplicate pool in AR when one is attached to a content host or activation key. Then any information about the pool relating the the ack key or content host can be updated in the relation table, like quantity attached. This will also allow attaching multiple groups of the same subscription to  a content host and have them be visible in different rows (satellite works this way now). 

Correct me if I am wrong, but I think we have a similar model for pools? 

This conversation can be taken elsewhere but I figured I would raise it here because this PR is a start to these changes. 

@ehelms @jlsherrill @thomasmckay

ehelms commented on app/controllers/katello/api/v2/subscriptions_controller.rb:
```diff
@@ -232,48 +218,29 @@ def default_sort
       %w(id desc)
     end
 
+    def index_subscriptions
```
This is nitpicky, but is this importing or indexing? Indexing makes me think of our ES setup, import our newer way of handling backend data.

ehelms commented on engines/bastion_katello/app/assets/javascripts/bastion_katello/subscriptions/subscriptionAttachAmountFilter.filter.js:
```diff
@@ -8,8 +8,16 @@
-        return function (subscription) {
-            var amount = subscription.amount;
+        var amount;
+        var i;
```
You can do this in one line or one var statement on two lines.

ehelms replied:
Actually, do these need to live outside the function using them in the return statement below?

ehelms commented on app/controllers/katello/api/v2/subscriptions_controller.rb:
```diff
@@ -81,24 +81,18 @@ def create
+        params[:subscriptions].each do |sub_params|
+          subscription = Pool.find(sub_params[:id])
+          object.subscribe(subscription.cp_id, sub_params[:quantity])
+          subscription.import_data
```
If I read this correctly, now when we do an attach of a subscription to an object, we are doing an attach which calls out to Candlepin and then doing a data import from Candlepin to ensure out database is correct w.r.t. the Candlepin data?

johnpmitsch replied:
@ehelms yes basically.The import_data is mostly to update the quantities consumed.

ehelms replied:
Since we are adding complexity, with an additional call out to Candlepin, I am thinking this should be dynflowized to better deal with potential errors. @jlsherrill would you agree?

johnpmitsch replied:
@ehelms I think he suggested this originally, is there a good example I can follow?

ehelms replied:
You can always pick a resource that you know is dynflowed (e.g. product creation) and try and follow that. There are also a number of recordings that cover Dynflow and conversion: https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&es_th=1&ie=UTF-8#q=dynflow&tbm=vid

I think those are two good places to start and then we can help fill in the gaps.

johnpmitsch replied:
thanks for sharing those, I'll look into it and make the changes here.

ehelms commented on app/models/katello/glue/candlepin/activation_key.rb:
```diff
@@ -37,12 +37,15 @@ def get_key_pools
         pools.each do |pool|
-          Katello::PoolActivationKey.where(:pool_id => pool.id, :activation_key_id => self.id).first_or_create
+          pool_activation_key = Katello::PoolActivationKey.where(:pool_id => pool.id, :activation_key_id => self.id).first_or_create
+          quantity = key_pools.collect { |x| x["amount"] if x["id"] == pool.cp_id }.compact.first if key_pools.any?
```
There is a lot of calls and logic going into this statement -- could we break it down a bit to reduce the complexity when reading it?

ehelms commented on app/models/katello/glue/candlepin/activation_key.rb:
```diff
@@ -37,12 +37,15 @@ def get_key_pools
+          pool_activation_key = Katello::PoolActivationKey.where(:pool_id => pool.id, :activation_key_id => self.id).first_or_create
+          quantity = key_pools.collect { |x| x["amount"] if x["id"] == pool.cp_id }.compact.first if key_pools.any?
+          pool_activation_key.update_attributes(:quantity => quantity) if quantity
+          pool_activation_key.save
```
Do we need to save this if the quantity isn't being updated?

ehelms commented on app/models/katello/glue/candlepin/consumer.rb:
```diff
@@ -88,6 +88,25 @@ def checkin(checkin_time)
       end
 
+      def pool_quantities
+        quantity_table = {}
```
Table is a confusing term to use here to me given it generally equates to a database table

ehelms commented on app/models/katello/glue/candlepin/consumer.rb:
```diff
@@ -88,6 +88,25 @@ def checkin(checkin_time)
+        pool_quantities = []
+        quantity_table.each do |key, value|
+          relation = {:pool_id => key.to_s.to_i, :quantity => value}
+          pool_quantities << relation
```
You could combine this statement and the above

ehelms commented on app/models/katello/glue/candlepin/consumer.rb:
```diff
@@ -88,6 +88,25 @@ def checkin(checkin_time)
+        quantity_table = {}
+        self.entitlements.each do |entitlement|
+          pool = self.get_pool(entitlement['pool']['id'])
+          pool_id = ::Katello::Pool.find_by_cp_id(pool["id"]).id.to_s.to_sym
```
Is all of this casting from one thing to another required?

ehelms commented on app/models/katello/glue/candlepin/pool.rb:
```diff
@@ -27,8 +27,7 @@ def candlepin_data(cp_id)
 
-      def get_for_owner(organization)
-        Katello::Resources::Candlepin::Pool.get_for_owner(organization)
+      def get_for_owner(organization) Katello::Resources::Candlepin::Pool.get_for_owner(organization)
```
Typo? This line on the same one as the `def` seems odd

ehelms commented on app/views/katello/api/v2/subscriptions/base.json.rabl:
```diff
@@ -13,3 +13,12 @@ attributes :sockets, :cores, :ram
+    id: subscription.host.uuid,
+    name: subscription.host.name
+  }
+end
```
Why do we need this information in the base rabl?

johnpmitsch replied:
https://github.com/Katello/katello/blob/068062449b30e985695aa6071837290e4afbd3df/engines/bastion_katello/app/assets/javascripts/bastion_katello/subscriptions/views/subscription-type.html#L5 this uses it to show subscription type and guest host subscriptions in the index

ehelms commented on app/views/katello/api/v2/subscriptions/base.json.rabl:
```diff
@@ -13,3 +13,12 @@ attributes :sockets, :cores, :ram
 attributes :type
 attributes :name => :product_name
+attributes :unmapped_guest
+attributes :virt_only
```
Should we only provide these attributes if it is virtual?

johnpmitsch replied:
@ehelms I am not sure if that will work with the checks here, since virt_only is being checked for false? http://git.io/vW5mC

ehelms replied:
I think you mean the other way around? Since it is checking explicitly for true, then that piece of UI should be fine. However, the UI shouldn't inherently dictate the API.

johnpmitsch replied:
@ehelms wrong line, my bad, I actually meant this one http://git.io/vW53Z Do  you think we only return virt_only and unmapped_guest if virtual and update the UI there?

ehelms replied:
Way to turn my question on me :) I was more looking for your opinion -- my question being one of are these attributes that make sense in most use cases or a very select few of cases?

johnpmitsch replied:
haha I wasn't sure if you had an opinion either way :) My opinion is to leave virt_only in because it is more explicit and can be used in logic easier when people use the api or in UI/hammer. Let me know if I am off base there.

ehelms replied:
:+1:

jlsherrill commented on engines/bastion_katello/app/assets/javascripts/bastion_katello/subscriptions/subscriptionAttachAmountFilter.filter.js:
```diff
@@ -8,8 +8,16 @@
+                if (attachedSubscriptions[i].pool_id === subscription.id) {
+                    amount = attachedSubscriptions[i].quantity;
+                }
+            }
```
we typically prefer using underscore js's  `_.each`  function.  Should be a number of examples in the code base

jlsherrill replied:
saves you from initializing 'i' as well :)

ehelms replied:
or angular.forEach since its native to the angular library

jlsherrill replied:
err yes, or that :)

ehelms commented:
Is there an issue for the proposal in your comment? I not quite following what the issue is since you mentioned "remaining issues" but don't quantify them.

johnpmitsch commented:
@ehelms this is one http://projects.theforeman.org/issues/12031 and the other (which still needs an issue filed) is not showing multiple rows of the same subscription attached to the content host. So if you attach 4 of subscription foo and then 4 more of subscription foo (same pool as well) it should show in multiple rows, which is how candlepin returns them and satellite does this today.

ehelms commented on app/views/katello/api/v2/subscriptions/base.json.rabl:
```diff
@@ -13,3 +13,12 @@ attributes :sockets, :cores, :ram
 attributes :instance_multiplier, :stacking_id, :multi_entitlement
 attributes :type
 attributes :name => :product_name
+attributes :unmapped_guest
```
What does this attribute actually mean? I looked through the code, and it appears from Candlepin it comes in as `unmapped_guests_only` which has a different meaning than `unmapped_guest` to me.

johnpmitsch replied:
I was just following the previous lazy accessor attribute here http://git.io/vW59b (from pre subscriptions scoped search) I can update the column to unmapped_guests_only and change where it is used. What about the API though?

ehelms replied:
Ahh right, sorry, missed that they were moving from show to base. We can leave them be since we've established this pattern of changing the attributes :)

ehelms commented on app/controllers/katello/api/v2/subscriptions_controller.rb:
```diff
@@ -17,8 +17,6 @@ class Api::V2::SubscriptionsController < Api::V2::ApiController
 
     api :GET, "/organizations/:organization_id/subscriptions", N_("List organization subscriptions")
-    api :GET, "/systems/:system_id/subscriptions", N_("List a content host's subscriptions"), :deprecated => true
-    api :GET, "/activation_keys/:activation_key_id/subscriptions", N_("List an activation key's subscriptions")
```
This would seem to be moving away from the pattern that we have been converging towards which is to use the index method of a resource to handle these types of queries. Is there a technical reason to make an exception here?

johnpmitsch replied:
Showing subscriptions for systems and activation keys require calling candlepin directly and returning what it sends in the API so we can have the quantity attached to a system or activation key. This wouldn't' return Pool (or Subscription) objects from the database but rather a presenter (SystemSubscriptionPresenter  or ActivationKeySubscriptionPresenter) and the subscription index call would be returning 2 different types of objects, which is further complicated by the index call using scoped search. I felt it would be simpler to move the calls to the system and activation key controllers.

ehelms replied:
How do the changes being made to system for scoped search conversion play into this? Do they change the workflow of how this data is collected at all? I know we've chatted about this prior, but it's all a big jumble in my head.

johnpmitsch replied:
If you mean how data is collected from candlepin, the changes with systems scoped search shouldn't affect that.

ehelms commented on engines/bastion_katello/app/assets/javascripts/bastion_katello/subscriptions/subscriptionAttachAmountFilter.filter.js:
```diff
@@ -9,7 +9,7 @@ angular.module('Bastion.subscriptions').filter('subscriptionAttachAmountFilter',
     function (translate) {
         return function (subscription) {
-            var amount = subscription.amount;
+            var amount = subscription.quantity_attached;
```
You will want to use hash access `[]` since this isn't camel case.

johnpmitsch commented:
@jlsherrill updated

cfouant commented:
@johnpmitsch - please update this p-r to handle updating subscription info when unregistering a system as well.

cfouant commented:
@johnpmitsch - Reproduction steps:

1) register a content host, and attach a subscription to it - note updated subscription quantity, e.g 1 out of 20
2) unregister content host `subscription-manager unregister`
3) refresh page to see subscription quantity - should now be 0 out of 20 (or however many less were associated with that content host)

This doesn't seem to be the case, although it does seem to update if you go to Red Hat Subscriptions > Manage Manifest > Refresh Manifest - that should not be necessary however.

johnpmitsch commented:
@cfouant updated system destroy action to import_all pools on finalization https://github.com/johnpmitsch/katello/blob/subscription-fixes/app/lib/actions/candlepin/consumer/destroy.rb#L14 which will update all pool/subscription information.

cfouant commented on test/controllers/api/v2/systems_controller_test.rb:
```diff
@@ -244,5 +247,21 @@ def test_product_content
       assert_template 'api/v2/systems/product_content'
     end
+
+    def test_add_subscriptions
```
could you rename this test `test_add_subscriptions_protected` ? APJ from me pending this line change. Great job!

johnpmitsch commented:
@cfouant good catch, updated. I'll wait for @jlsherrill to comment before merging

cfouant commented:
@johnpmitsch yes, definitely wait on @jlsherrill !!!

cfouant commented:
@johnpmitsch - I think you'll need to add `:add_subscriptions` to [content_host_permissions](../lib/katello/permissions/content_host_permissions.rb) (line 22)

Also, the test should look like this:

`def test_add_subscriptions_protected
    allowed_perms = [@update_permission]
    denied_perms = [@view_permission, @create_permission, @destroy_permission]

    assert_protected_action(:add_subscriptions, allowed_perms, denied_perms) do
      post :add_subscriptions, :id => @system.uuid, :subscriptions => [{:id => 'redhat', :quantity => 1}]
    end
end'

johnpmitsch commented:
@cfouant cool, thanks, must have ignored that one when it was initially passing. I've updated the test_add_subscriptions_protected test

jlsherrill commented on app/lib/actions/candlepin/consumer/destroy.rb:
```diff
@@ -9,6 +9,10 @@ class Destroy < Candlepin::Abstract
+
+        def finalize
+          ::Katello::Pool.import_all
+        end
```
This really belongs in a different action, such as  katello/system/destroy.rb

also, reindexing all pools across all orgs probably isn't ideal.  We should try to limit this to a single org, or potentially just the pools the system was consuming?

johnpmitsch commented:
@cfouant @jlsherrill updated to just import data on the pools subscriptions (not all of them) in https://github.com/johnpmitsch/katello/blob/subscription-fixes/app/lib/actions/katello/system/destroy.rb

cfouant commented:
NACK - I have to take my APJ back... I'm finding that if you register and auto-attach using a RHEL vm (either with key or direct registration), subscriptions are not updating. This isn't a problem for Centos vms because you have to manually add the subscriptions to them. It doesn't seem like `Katello::Pool.find_by_cp_id(entitlement["pool"]["id"])` is working because the pool is not being found in the database and returning nil.

johnpmitsch commented:
@jlsherrill @cfouant updated with my proposed solution, I have changed https://github.com/Katello/katello/pull/5536/files#diff-d4ada19367f2e37f2adeafa9de778bccR530 to take an argument called include_temporary_guests which defaults to false. Its passed true for an argument when get_for_owner is called for ::Katello::Pool, which imports the temporary subscription when import_all is called. In the index of subscription controller, I changed it to only return when unmapped_guest column is false, which is now default in the migration (you'll have to reset or migrate) https://github.com/Katello/katello/pull/5536/files#diff-937e4184d63313d2975db7d87ebd1c90R37 This way a temporary subscription will not show up in an index call, but it is in the database when a content host is finding Pools by id.

cfouant commented:
@johnpmitsch - I'm still not seeing the subscription amounts updating upon registering a RHEL machine (either by activation key or with username/password and --auto-attach flag). Is the proposed solution in this yet, or are you just asking for feedback before implementing?

johnpmitsch commented:
@cfouant the proposed solution is in the code, I haven't tested it as thoroughly as I should but pushed it up for initial feedback in case I am on the wrong track and have to rewrite that part. I'll take a look tomorrow and get back to you on it, but if I remember correctly, the temporary subscription is a separate pool and was showing up in the list/remove part of the content host after auto-attach.
As far as subscriptions showing up in the subscriptions table (on an index call) did you rake katello:reset? The migrations have changed and that will affect what shows up on an index call.

johnpmitsch commented:
@cfouant I got a chance to look at this more,I am seeing the temporary subscription is not updating its quantity. There was '0 out of unlimited' on the temporary subscription and when I ran import_data on that subscription in the rails console it was updated to '1 out of unlimited' like here http://www.clipular.com/c/5697988432232448.png?k=FTT8Qz2b9rktG8neCysa5KukWYo, I'll fix and update. However, I am not seeing temporary subscriptions showing up on an index call (like in subscriptions main page), if you continue to see this after a reset, can you let me know if any of the Katello::Pool's have a :virt_only => true value?

cfouant commented:
@johnpmitsch - I am going to check again with a reset database, but where I saw it was on the content host details Subscriptions tab, when you click 'add'. They have the temporary subscriptions listed in subscriptions to add to a content host. Possibly same on an activation key, but can't recall.

johnpmitsch commented:
@cfouant I thought temporary subscriptions should be shown to add to a content host? @jlsherrill @thomasmckay  is this correct?

cfouant commented:
@jlsherrill @thomasmckay @johnpmitsch - Could be the case. I was under the impression that the temporary subscription is added automatically and behind the scenes, and that we wanted the functionality of an active subscription, but still hidden. Getting confused though.

johnpmitsch commented:
@cfouant If I understand @thomasmckay's explanation correctly, temporary subscriptions should be visible to a content host.

cfouant commented:
@thomasmckay - can you confirm? I would understand why we would want to see temporary subscriptions that exist on a content host, but would we want them to be available for adding manually to a content host, or via activation key?

thomasmckay commented:
@cfouant - Yes, they should be listed to manually attach to content host. Yes, the should be shown when attached to content host. No, they should not be listed on Subscriptions page. No, they should not be listed anywhere on Activation Keys.

johnpmitsch commented:
@cfouant what I am seeing locally -- temporary subscriptions are showing where they should, but the quantity on the temporary subscription is not updating after an auto-attach. Is this in line with what you are seeing?

cfouant commented:
@johnpmitsch - yes. Non temporary subscriptions quantities are also not updating on rhel machines.

johnpmitsch commented:
@cfouant @jlsherrill @thomasmckay Fixing updating temporary subscription quantities is currently being blocked by http://projects.theforeman.org/issues/12475 as the auto-attach happens in candlepin, and to update the subscription quantities for temporary subscriptions would require listening for a message from candlepin via qpid, which is not working correctly in a dev environment.

If my understanding of all this is correct, then it would appear we have two options:
1) Fix any non-qpid-related issues in this PR and then merge after review, then open up issue for temporary subscriptions not being updated and assign to me.
2) Wait for 12475 to be resolved before updating, reviewing, and merging this PR.

I am in favor of option 1, but want to hear your thoughts/opinions

thomasmckay commented:
I am in favor of merging this ASAP and dealing with any further issues. As it stands right now, subscriptions are all but unusable in master.

cfouant commented:
@johnpmitsch I'm with @thomasmckay on this one. I would link the new issue you open in this PR as well.

johnpmitsch commented:
@cfouant @thomasmckay sounds good, someone go ahead and ACK and I will merge

cfouant commented:
ACK from me, but please hold on merge until getting ACKs from the other commenters in this thread @jlsherrill @thomasmckay

thomasmckay commented:
@johnpmitsch Mind rebasing this against master? I will test again today with the puppet-katello PRs that fix the qpid events from candlepin.

johnpmitsch commented:
@thomasmckay rebased

thomasmckay commented:
systems_controller.rb:
+ GET /subscriptions - missing pagination apipie params
+ DELETE /subscriptions - no such method

thomasmckay commented:
ACK - There are remaining issues in the area of subscriptions but they can be addressed individually outside of this PR. I will defer to @jlsherrill to give final ACK to merge.

jlsherrill commented:
ACK
